packages/node-patches/src/fs.ts

```diff
@@ -0,0 +1,345 @@
+  fs.lstatSync = (...args: any[]) => {
+    let stats = origLstatSync(...args);
+    if (stats.isSymbolicLink() && isOutPath(origRealpathSync(args[0]))) {
+      stats = fs.statSync(...args);
+    }
+    return stats;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.realpathSync = (...args: any[]) => {
+    const str = origRealpathSync(...args);
+    if (isOutPath(str)) {
+      return path.resolve(args[0]);
+    }
+    return str;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readlinkSync = (...args: any[]) => {
+    const str = origReadlinkSync(...args);
+    if (isOutPath(str)) {
+      const e = new Error(
+        "EINVAL: invalid argument, readlink '" + args[0] + "'"
+      );
+      //tslint:disable-next-line:no-any
+      (e as any).code = 'EINVAL';
+      throw e;
+    }
+    return str;
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readdir = (...args: any[]) => {
+    const p = path.resolve(args[0]);
+
+    const cb = args[args.length - 1];
+    if (typeof cb !== 'function') {
+      // this will likely throw callback required error.
+      return origReaddir(...args);
+    }
+
+    args[args.length - 1] = (err: Error, result: Dirent[]) => {
+      if (err) return cb(err);
+      // user requested withFileTypes
+      if (result[0] && result[0].isSymbolicLink) {
+        Promise.all(result.map((v: Dirent) => handleDirent(p, v)))
+          .then(() => {
+            cb(null, result);
+          })
+          .catch(err => {
+            cb(err);
+          });
+      } else {
+        // string array return for readdir.
+        cb(null, result);
+      }
+    };
+
+    origReaddir(...args);
+  };
+
+  //tslint:disable-next-line:no-any
+  fs.readdirSync = (...args: any[]) => {
+    const res = origReaddirSync(...args);
+    const p = path.resolve(args[0]);
+    //tslint:disable-next-line:no-any
+    res.forEach((v: Dirent | any) => {
+      handleDirentSync(p, v);
+    });
+    return res;
+  };
+
+  // i need to use this twice in bodt readdor and readdirSync. maybe in fs.Dir
+  //tslint:disable-next-line:no-any
+  function patchDirent(dirent: Dirent | any, stat: Stats | any) {
+    // add all stat is methods to Dirent instances with their result.
+    for (const i in stat) {
+      if (i.indexOf('is') === 0 && typeof stat[i] === 'function') {
+        const result = stat[i]();
+        if (result) dirent[i] = () => true;
+        else dirent[i] = () => false;
+      }
+    }
+  }
+
+  if (fs.opendir) {
+    const origOpendir = fs.opendir.bind(fs);
+    //tslint:disable-next-line:no-any
+    fs.opendir = (...args: any[]) => {
+      const cb = args[args.length - 1];
+      // if this is not a function opendir should throw an error.
+      //we call it so we don't have to throw a mock
+      if (typeof cb === 'function') {
+        args[args.length - 1] = async (err: Error, dir: Dir) => {
+          try {
+            cb(null, await handleDir(dir));
+          } catch (e) {
+            cb(e);
```

Comment:
Similar to above: Is there a risk that we call `cb` twice if it throws?

Reply:
oh yeah if the execution of the cb itself throws. yes if course